SU2_CFD/src/driver_structure.cpp

```diff
@@ -3081,37 +3209,139 @@ void CSingleZoneDriver::SetInitialMesh() {
+CFluidDriver::CFluidDriver(char* confFile, unsigned short val_nZone, unsigned short val_nDim) : CDriver(confFile, val_nZone, val_nDim) { }
 
+CFluidDriver::~CFluidDriver(void) { }
 
-CMultiZoneDriver::~CMultiZoneDriver(void) { }
-
-void CMultiZoneDriver::Run() {
-  
+void CFluidDriver::Run() {
   
+  unsigned short iZone, jZone, checkConvergence;
+  unsigned long IntIter, nIntIter;
+  bool unsteady;
+
   /*--- Run a single iteration of a multi-zone problem by looping over all
-   zones and executing the iterations. Note that data transers between zones
-   and other intermediate procedures may be required. ---*/
-  
-  for (iZone = 0; iZone < nZone; iZone++) {
-    
-    iteration_container[iZone]->Preprocess(output, integration_container, geometry_container,
-                                           solver_container, numerics_container, config_container,
-                                           surface_movement, grid_movement, FFDBox, iZone);
+  zones and executing the iterations. Note that data transers between zones
+  and other intermediate procedures may be required. ---*/
+
+  unsteady = (config_container[MESH_0]->GetUnsteady_Simulation() == DT_STEPPING_1ST) || (config_container[MESH_0]->GetUnsteady_Simulation() == DT_STEPPING_2ND);
+
+  /*--- Zone preprocessing ---*/
+
+  for (iZone = 0; iZone < nZone; iZone++)
+  	iteration_container[iZone]->Preprocess(output, integration_container, geometry_container, solver_container, numerics_container, config_container, surface_movement, grid_movement, FFDBox, iZone);
+
+  /*--- Updating zone interface communication patterns,
+  needed only for unsteady simulation since for steady problems
+  this is done once in the interpolator_container constructor 
+  at the beginning of the computation ---*/
+
+  if ( unsteady ){
+    for (iZone = 0; iZone < nZone; iZone++) {   
+  	  for (jZone = 0; jZone < nZone; jZone++)
+  	    if(jZone != iZone && interpolator_container[iZone][jZone] != NULL)
+  		  interpolator_container[iZone][jZone]->Set_TransferCoeff(config_container);
+	}
+  }
+
+
+  /*--- Begin Unsteady pseudo-time stepping internal loop, if not unsteady it does only one step --*/
+
+  if (unsteady) 
+    nIntIter = config_container[MESH_0]->GetUnst_nIntIter();
+  else
+    nIntIter = 1;
+
+  for (IntIter = 0; IntIter < nIntIter; IntIter++){
+
+    /*--- At each pseudo time-step updates transfer data ---*/
+    for (iZone = 0; iZone < nZone; iZone++)   
+	  for (jZone = 0; jZone < nZone; jZone++)
+	    if(jZone != iZone && transfer_container[iZone][jZone] != NULL)
+		  Transfer_Data(iZone, jZone);
+
+	/*--- For each zone runs one single iteration ---*/
+	for (iZone = 0; iZone < nZone; iZone++) {
+      config_container[iZone]->SetIntIter(IntIter);
+
+      iteration_container[iZone]->Iterate(output, integration_container, geometry_container, solver_container, numerics_container, config_container, surface_movement, grid_movement, FFDBox, iZone);
+	}
```

Comment:
Hi,
I can see that this may be a problem for FSI applications. The old implementation was indeed correct for time domain simulations as the iterations need to be sequential. 

The coupling between zones for FSI has to be accounted only at physical time-steps, it is not reasonable in that context to transfer every pseudo step between the subiterations of the flow and the structure.

I would need some time to further review that this does not affect my previous work.